src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs

```diff
@@ -175,22 +206,115 @@ private ITextBuffer TryGetTextBufferFromDocData(object docData)
             }
         }
 
-        private void NewBufferOpened(uint docCookie, ITextBuffer textBuffer, DocumentKey documentKey, bool isCurrentContext)
+        private void NewBufferOpened(uint docCookie, ITextBuffer textBuffer, DocumentKey documentKey, bool isCurrentContext, CancellationToken cancellationToken)
         {
+            AssertIsForeground();
+
+            cancellationToken.ThrowIfCancellationRequested();
+
             StandardTextDocument document;
+            lock (_gate)
+            {
+                if (!_documentMap.TryGetValue(documentKey, out document))
+                {
+                    return;
+                }
+            }
 
-            if (_documentMap.TryGetValue(documentKey, out document))
+            document.ProcessOpen(textBuffer, isCurrentContext);
+
+            lock (_gate)
             {
-                document.ProcessOpen(textBuffer, isCurrentContext);
-                AddCookieOpenDocumentPair(docCookie, documentKey);
+                AddCookieOpenDocumentPair_NoLock(docCookie, documentKey);
             }
         }
 
         /// <summary>
         /// Notifies the document provider that this document is now registered in a project.
+        /// If we are on a foregroud thread, then this is done right away.
+        /// Otherwise, we schedule a task on foreground task scheduler.
         /// </summary>
         public void NotifyDocumentRegisteredToProject(IVisualStudioHostDocument document)
```

Comment:
Maybe rename this to NotifyDocumentRegisteredToProjectAndStartToRaiseEvents or something? I don't mind it's doing two things, but the contract is clearer.